libvast/vast/format/json.hpp

```diff
@@ -41,4 +53,202 @@ class writer : public printer_writer<event_printer>{
   }
 };
 
+struct default_selector {
+  caf::optional<vast::record_type> operator()(const vast::json::object&) {
+    return t;
+  }
+
+  caf::error schema(vast::schema sch) {
+    auto entry = *sch.begin();
+    if (!caf::holds_alternative<vast::record_type>(entry))
+      return make_error(vast::ec::invalid_configuration,
+                        "only record_types supported for json schema");
+    t = flatten(caf::get<vast::record_type>(entry));
+    return caf::none;
+  }
+
+  vast::schema schema() const {
+    vast::schema result;
+    if (t)
+      result.add(*t);
+    return result;
+  }
+
+  static const char* name() {
+    return "json-reader";
+  }
+
+  caf::optional<vast::record_type> t = caf::none;
+};
+
+template <class Selector = default_selector>
+class reader final : public multi_layout_reader {
+public:
+  using super = multi_layout_reader;
+
+  /// Constructs a json reader.
+  /// @param input The stream of logs to read.
+  explicit reader(caf::atom_value table_slice_type,
+                  std::unique_ptr<std::istream> in = nullptr);
+
+  void reset(std::unique_ptr<std::istream> in);
+
+  caf::error schema(vast::schema sch) override;
+
+  caf::error schema(vast::type, vast::schema);
+
+  vast::schema schema() const override;
+
+  const char* name() const override;
+
+protected:
+  caf::error read_impl(size_t max_events, size_t max_slice_size,
+                       consumer& f) override;
+
+private:
+  using iterator_type = std::string_view::const_iterator;
+
+  // void patch(std::vector<data>& xs);
+
+  Selector selector;
+  std::unique_ptr<std::istream> input_;
+  std::unique_ptr<vast::detail::line_range> lines_;
+  caf::optional<size_t> proto_field_;
+  std::vector<size_t> port_fields_;
+};
+
+template <class Selector>
+reader<Selector>::reader(caf::atom_value table_slice_type,
+                         std::unique_ptr<std::istream> in)
+  : super(table_slice_type) {
+  if (in != nullptr)
+    reset(std::move(in));
+}
+
+template <class Selector>
+void reader<Selector>::reset(std::unique_ptr<std::istream> in) {
+  VAST_ASSERT(in != nullptr);
+  input_ = std::move(in);
+  lines_ = std::make_unique<vast::detail::line_range>(*input_);
+}
+
+template <class Selector>
+caf::error reader<Selector>::schema(vast::schema s) {
+  return selector.schema(std::move(s));
+}
+
+template <class Selector>
+vast::schema reader<Selector>::schema() const {
+  return selector.schema();
+}
+
+template <class Selector>
+const char* reader<Selector>::name() const {
+  return Selector::name();
+}
+
+template <class F>
+struct convert {
+  convert(F&& f) : f_{std::forward<F>(f)} {
```

Comment:
```suggestion
  convert(F f) : f_{f} {
```

For functions, we could use this simpler form. But I'm also fine with the way it is right now.

Reply:
I'd prefer to keep the rvalue version, this way we don't need to worry about the size of the capture.